chore: upgrade node 14

coryrylan opened the pull request:
## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [ ] If applicable, have a visual design approval

## PR Type

What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [x] CI related changes
- [ ] Documentation content changes
- [ ] clarity.design website / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

From the Big Sur update it looks like the angular ngcc compile step can cause a file lock to time out. This prevents the repo form building.

```
Compiling @angular/router : esm2015 as esm2015
Compiling @angular/router : main as umd
Compiling @angular/router/testing : esm2015 as esm2015
Compiling @angular/router/testing : main as umd
events.js:292
      throw er; // Unhandled 'error' event
      ^
Error: kill EPERM
    at ChildProcess.kill (internal/child_process.js:492:26)
    at Worker.<anonymous> (internal/cluster/master.js:367:40)
    at Object.onceWrapper (events.js:421:28)
    at Worker.emit (events.js:315:20)
    at ChildProcess.<anonymous> (internal/cluster/master.js:212:12)
    at Object.onceWrapper (events.js:421:28)
    at ChildProcess.emit (events.js:315:20)
    at finish (internal/child_process.js:861:14)
    at processTicksAndRejections (internal/process/task_queues.js:79:11)
Emitted 'error' event on Worker instance at:
    at ChildProcess.<anonymous> (internal/cluster/worker.js:29:12)
    at ChildProcess.emit (events.js:315:20)
    at ChildProcess.kill (internal/child_process.js:492:12)
    at Worker.<anonymous> (internal/cluster/master.js:367:40)
    [... lines matching original stack trace ...]
    at finish (internal/child_process.js:861:14) {
  errno: 'EPERM',
  code: 'EPERM',
```

Issue Number: N/A

## What is the new behavior?
This updates the repo to Node 14. This seems to improve the issue but I consistently get the same ngcc issue the first time running the angular unit tests. After the first time, subsequent builds work fine. If I remove node_modules and re-install I will get the same error the first attempt only. 

I don't know what the ideal fix is. I have been stuck on this for a while but wanted to put this up for others to pull down and try.

blocked on https://github.com/vmware/clarity/pull/5300 merging 

closes https://github.com/vmware/clarity/issues/5696 

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

sis0k0 commented on packages/angular/projects/cds-angular/package.json:
```diff
@@ -8,8 +8,8 @@
   "peerDependencies": {
-    "@angular/common": "^9.0.0",
-    "@angular/core": "^9.0.0",
+    "@angular/common": "^11.0.0",
```
That way, installing the package in a project that uses Angular 9 or 10 will show a warning that `@cds/angular` requires Angular 11. Shouldn't the package be fine with these versions?

coryrylan replied:
Yeah, good catch. I'll get it updated for `>9` I think I had moved it to 11 because of yarn not playing well with the multiple angular versions.

sis0k0 replied:
Yes, if we want it to respect 9, 10 and 11, we should set it to
```suggestion
    "@angular/common": "9.x || 10.x || 11.x",
```

sis0k0 commented on packages/angular/projects/cds-angular/package.json:
```diff
@@ -8,8 +8,8 @@
-    "@angular/common": "^9.0.0",
-    "@angular/core": "^9.0.0",
+    "@angular/common": "^11.0.0",
+    "@angular/core": "^11.0.0",
```
Same as the line above:
```suggestion
    "@angular/core": "9.x || 10.x || 11.x",
```

coryrylan committed:
chore: update to node 14
- update to node 14
- fix circular dependencies related to update
- fix karma timeouts
- remove deprecated ngcc calls

Signed-off-by: Cory Rylan <splintercode.cb@gmail.com>

coryrylan commented on apps/core-angular-universal/package.json:
```diff
@@ -10,8 +10,7 @@
     "dev:ssr": "ng run core-angular-universal:serve-ssr",
     "serve:ssr": "node dist/core-angular-universal/server/main.js",
-    "prerender": "ng run core-angular-universal:prerender",
-    "postinstall": "ngcc && node ../../scripts/ngcc-clean"
```
ngcc is no longer required or should be called in latest angular

coryrylan commented on apps/website/.vuepress/theme/global-components/Badge.vue:
```diff
@@ -27,22 +27,32 @@ export default {
+
+.badge + .badge {
+  margin-left: 5px;
+}
```
Converted this to scss from stylus. This was the only place stylus was used which required the package that was causing issues in node 14 https://github.com/stylus/stylus/issues/2534

coryrylan commented on packages/angular/package.json:
```diff
@@ -13,15 +13,12 @@
-    "build:package:clr": "replace '@VERSION' $npm_package_version ./dist/clr-angular/package.json",
-    "build:package:cds": "replace '@VERSION' $npm_package_version ./dist/cds-angular/package.json",
+    "build:preinstall:cds": "mkdir -p ../../dist/angular/ && cp ./projects/cds-angular/package.json ../../dist/angular/",
+    "build:preinstall:clr": "mkdir -p ../../dist/clr-angular/ && cp ./projects/clr-angular/package.json ../../dist/clr-angular/",
```
Updated preinstall to copy the package json like the other packages. This ensure yarn installs their dependencies at the top level of the mono repo. Without it yarn updates/installs will fail because the angular package dependencies were never installed.

coryrylan commented on packages/angular/package.json:
```diff
@@ -53,6 +50,8 @@
     "@angular/router": "11.0.0",
     "@cds/core": "./dist/core",
+    "@cds/angular": "./dist/angular",
+    "@clr/angular": "./dist/clr-angular",
```
This ensures the dev apps in the angular project get the local copies when built.

coryrylan commented on packages/angular/package.json:
```diff
@@ -104,7 +103,7 @@
     "lit-element": "2.3.1",
     "lit-html": "1.2.1",
-    "ng-packagr": "11.0.2",
+    "ng-packagr": "11.2.4",
```
Had to bump for node 14, get around the stylus bug mentioned above,

coryrylan commented on packages/angular/projects/clarity-adoption/tsconfig.app.json:
```diff
@@ -3,7 +3,13 @@
+      "@clr/angular/*": ["dist/clr-angular/*"],
+      "@cds/angular": ["dist/angular"],
+      "@cds/angular/*": ["dist/angular/*"]
+    }
```
Map the dist output so when the source code changes in the lib the dev apps auto reload. The mapping should only be at the demo app level. When listed at the top level it creates circular references causing excessive build times. Related: https://github.com/angular/angular/issues/38832

coryrylan commented on packages/core/package.json:
```diff
@@ -86,12 +86,12 @@
-    "karma-jasmine": "2.0.1",
-    "karma-jasmine-html-reporter": "1.4.2",
+    "karma-jasmine": "4.0.1",
+    "karma-jasmine-html-reporter": "1.5.4",
```
Bump for node 14 and fixes the test timeouts we are currently seeing

coryrylan commented on packages/core/src/internal/utils/identity.spec.ts:
```diff
@@ -339,4 +340,37 @@ describe('Functional Helper: ', () => {
       expect(getMillisecondsFromSeconds(-0.5)).toBe(-500);
     });
   });
+
```
I had to swap the function over as it was creating a circular dependency between the identity and string utils.

coryrylan commented on packages/icons/package.json:
```diff
@@ -1,15 +1,15 @@
     "build:css": "sass --precision=8 src/clr-icons.scss ./dist/clr-icons/clr-icons.css",
     "build:optimize": "csso -i ./dist/clr-icons/clr-icons.css -o ./dist/clr-icons/clr-icons.min.css -s file --no-restructure;",
-    "build:package": "cpy npm.json README.md ./dist/clr-icons/; mv ./dist/clr-icons/npm.json ./dist/clr-icons/package.json; replace '@VERSION' $npm_package_version ./dist/clr-icons/package.json; ",
+    "build:package": "cpy npm.json README.md ./dist/clr-icons/; mv ./dist/clr-icons/npm.json ./dist/clr-icons/package.json;",
```
Updated to be consistent with the rest of the repo and use the version copy script

coryrylan commented on packages/ui/package.json:
```diff
@@ -12,7 +12,7 @@
     "build:prefix": "postcss ./dist/clr-ui/clr-ui.css -o ./dist/clr-ui/clr-ui.css; postcss ./dist/clr-ui/clr-ui-dark.css -o ./dist/clr-ui/clr-ui-dark.css",
     "build:optimize": "csso -i ./dist/clr-ui/clr-ui.css -o ./dist/clr-ui/clr-ui.min.css -s file --no-restructure; csso -i ./dist/clr-ui/clr-ui-dark.css -o ./dist/clr-ui/clr-ui-dark.min.css -s file --no-restructure",
-    "build:package": "cpy npm.json README.md ./dist/clr-ui/; mv ./dist/clr-ui/npm.json ./dist/clr-ui/package.json; replace '@VERSION' $npm_package_version ./dist/clr-ui/package.json; "
+    "build:package": "cpy npm.json README.md ./dist/clr-ui/; mv ./dist/clr-ui/npm.json ./dist/clr-ui/package.json;"
```
Updated to be consistent with the rest of the repo and use the version copy script

coryrylan commented on scripts/copy-version.js:
```diff
@@ -13,12 +13,13 @@ const path = require('path');
   'packages/eslint-plugin-clarity-adoption/package.json',
-  'packages/ui/package.json',
+  'packages/icons/npm.json',
+  'packages/ui/npm.json',
```
Updated packages to all use the same version copy util. Fixes yarn install issues an build inconsistency
